pkg/network/stats.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"time"
+)
+
+// ReqEvent represents either an incoming or closed request.
+// +k8s:deepcopy-gen=false
+type ReqEvent struct {
+	Time time.Time
+	Type ReqEventType
+}
+
+// ReqEventType denotes the type (incoming/closed) of a ReqEvent.
+type ReqEventType int
+
+const (
+	// ReqIn represents an incoming request
+	ReqIn ReqEventType = iota
+	// ReqOut represents a finished request
+	ReqOut
+	// ProxiedIn represents an incoming request through a proxy.
+	ProxiedIn
+	// ProxiedOut represents a finished proxied request.
+	ProxiedOut
+)
+
+// NewRequestStats builds a RequestStats instance, started at the given time.
+func NewRequestStats(startedAt time.Time) *RequestStats {
+	return &RequestStats{lastChange: startedAt}
+}
+
+// RequestStats collects statistics about requests as they flow in and out of the system.
+// It's usually used in a flow where events are generated asynchronously and fed into
+// the object by reading from a channel. Report is likewise called from a different
+// channel so the pattern usually looks like:
+//
+// stats := NewRequestStats(time)
+// for {
+//   switch {
+//   case e <- reqChan:
+//     stats.HandleEvent(e)
+//   case now <- reportChan:
+//     stats.Report(now)
+//   }
+// }
+//
+// The individual functions are not thread-safe. They are safe to use in the
+// single-threaded pattern shown above.
+//
+// +k8s:deepcopy-gen=false
+type RequestStats struct {
+	// State variables that track the current state. Not reset after reporting.
+	concurrency, proxiedConcurrency float64
+	lastChange                      time.Time
+
+	// Reporting variables that track state over the current window. Reset after
+	// reporting.
+	requestCount, proxiedCount                      float64
```

Comment:
I'd be tempted to put all of these in to a WindowStats struct (or sthg like that). That way it's clear which instance vars have a different lifecycle than the rest without scrolling back up to this comment. Also that way we can lose most of the body of reset() and explicitly zero that struct, which avoids some potential bugs and is maybe a bit clearer about what's going on. This is a small-ish file though so may be overkill, lgtm apart from that now.